pkg/util/capacity/conditions.go

```diff
@@ -0,0 +1,126 @@
+	c1, c2 *shipper.ClusterCapacityCondition
+}
+
+var _ diff.Diff = (*ClusterCapacityConditionDiff)(nil)
+
+func NewClusterCapacityConditionDiff(c1, c2 *shipper.ClusterCapacityCondition) *ClusterCapacityConditionDiff {
+	return &ClusterCapacityConditionDiff{
+		c1: c1,
+		c2: c2,
+	}
+}
+
+func (d *ClusterCapacityConditionDiff) IsEmpty() bool {
+	if d.c1 == nil && d.c2 == nil {
+		return true
+	}
+	if d.c1 == nil || d.c2 == nil {
+		return false
+	}
+	return d.c1.Type == d.c2.Type &&
+		d.c1.Status == d.c2.Status &&
+		d.c1.Reason == d.c2.Reason &&
+		d.c1.Message == d.c2.Message
+}
+
+func (d *ClusterCapacityConditionDiff) String() string {
+	if d.IsEmpty() {
+		return ""
+	}
+	c1str, c2str := condStr(d.c1), condStr(d.c2)
+	return fmt.Sprintf("[%s] -> [%s]", c1str, c2str)
+}
+
+func NewClusterCapacityCondition(condType shipper.ClusterConditionType, status corev1.ConditionStatus, reason, message string) *shipper.ClusterCapacityCondition {
+	now := metav1.Now()
+	if CapacityConditionsShouldDiscardTimestamps {
+		now = metav1.Time{}
+	}
+	return &shipper.ClusterCapacityCondition{
+		Type:               condType,
+		Status:             status,
+		LastTransitionTime: now,
+		Reason:             reason,
+		Message:            message,
+	}
+}
+
+func SetClusterCapacityCondition(status *shipper.ClusterCapacityStatus, condition shipper.ClusterCapacityCondition) diff.Diff {
+	currentCond := GetClusterCapacityCondition(*status, condition.Type)
+
+	diff := NewClusterCapacityConditionDiff(currentCond, &condition)
+	if diff.IsEmpty() {
+		return nil
+	}
+
+	if currentCond != nil && currentCond.Status == condition.Status {
+		condition.LastTransitionTime = currentCond.LastTransitionTime
+	}
+
+	newConditions := filterOutCondition(status.Conditions, condition.Type)
+	status.Conditions = append(newConditions, condition)
+	sort.Slice(status.Conditions, func(i, j int) bool {
+		return status.Conditions[i].Type < status.Conditions[j].Type
+	})
+
+	return diff
+}
+
+func GetClusterCapacityCondition(status shipper.ClusterCapacityStatus, condType shipper.ClusterConditionType) *shipper.ClusterCapacityCondition {
+	for _, c := range status.Conditions {
+		if c.Type == condType {
+			return &c
+		}
+	}
+	return nil
+}
+
+func filterOutCondition(conditions []shipper.ClusterCapacityCondition, condType shipper.ClusterConditionType) []shipper.ClusterCapacityCondition {
+	var newConditions []shipper.ClusterCapacityCondition
+	for _, c := range conditions {
+		if c.Type == condType {
+			continue
+		}
+		newConditions = append(newConditions, c)
+	}
+	return newConditions
+}
+
+func condStr(c *shipper.ClusterCapacityCondition) string {
+	if c == nil {
+		return ""
+	}
+	chunks := []string{
+		fmt.Sprintf("%v", c.Type),
+		fmt.Sprintf("%v", c.Status),
+		c.Reason,
+		c.Message,
+	}
+	b := strings.Builder{}
+	for _, ch := range chunks {
+		if len(ch) > 0 {
+			if b.Len() > 0 {
+				b.WriteByte(' ')
+			}
+			b.WriteString(ch)
+		}
+	}
+	return b.String()
+}
```

Comment:
I'd say that with funcs that mostly deal with primitives like this one and `String()`, you could extract the logic to an util func that these just delegate to, to cut down on duplication a little bit.